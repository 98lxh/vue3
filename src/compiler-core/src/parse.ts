import { NodeTypes } from "./ast";


const enum TagType {
  Start,
  End
}

export function baseParse(content: string) {

  const context = createParseContext(content)
  return createRoot(parseChildren(context))
}

function parseChildren(context) {
  const nodes: any[] = [];
  let node
  const s = context.source
  if (s.startsWith('{{')) {
    node = parseInterpolation(context)
  } else if (s[0] === '<') {
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context)
    }
  }

  nodes.push(node)
  return nodes
}


function parseElement(context) {
  const element = parseTag(context, TagType.Start)

  parseTag(context, TagType.End)
  console.log(context.source)
  return element
}


function parseTag(context, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match[1]
  advance(context, match[0].length)
  advance(context, 1)

  if (type === TagType.End) return

  return {
    type: NodeTypes.ElEMENT,
    tag: tag
  }
}

function parseInterpolation(context) {

  const openDelimiter = '{{'
  const closeDelimiter = '}}'
  //ä¾‹:{{message}}
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

  //message}}
  advance(context, openDelimiter.length)

  const rawContentLength = closeIndex - openDelimiter.length
  //message
  const rawcontent = context.source.slice(0, rawContentLength)
  const content = rawcontent.trim()

  advance(context, rawContentLength + closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content
    }
  }
}

function advance(context: any, length: number) {
  context.source = context.source.slice(length)
}


function createRoot(children) {
  return {
    children
  }
}


function createParseContext(content: string) {
  return {
    source: content
  }
}
