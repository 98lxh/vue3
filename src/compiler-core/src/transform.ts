import { NodeTypes } from "./ast"

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  //遍历 深度优先搜索
  traversNode(root, context)

  createRootCodegen(root)
}


function createRootCodegen(root) {
  root.codegenNode = root.children[0]
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || []
  }
  return context
}

function traversNode(node, context) {
  const children = node.children

  const nodeTransforms = context.nodeTransforms

  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node)
  }

  traversChildren(children, context)
}

function traversChildren(children, context) {
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traversNode(node, context)
    }
  }
}
