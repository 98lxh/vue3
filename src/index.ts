export * from "./runtime-core/index";

//出口
