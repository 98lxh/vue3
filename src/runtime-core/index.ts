export { nextTick } from "./scheduler";

export { createRender } from "./renderer";
export { provide, inject } from "./apiInject";
export { getCurrentInstance } from "./component";
export { createTextVnode } from "./vnode";
export { renderSlots } from "./helper/renderSlots";
export { h } from "./h"
