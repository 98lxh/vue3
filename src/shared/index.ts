export const extend = Object.assign

export const hasChanged = (val, newVal) => {
  return !Object.is(val, newVal)
}

export const isObject = (val) => {
  return val !== null && typeof val === 'object'
}
