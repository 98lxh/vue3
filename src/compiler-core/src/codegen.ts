import { isString } from "../../shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context

  genFunctionPreamble(ast, context)

  const functionName = "render"
  const args = ["_ctx", "_cache"]
  const signature = args.join(",")

  push(`function ${functionName}(${signature}){`)
  push('return ')
  genNode(ast.codegenNode, context)
  push('}')

  return {
    code: context.code
  }
}

function genFunctionPreamble(ast, context) {
  const { push } = context
  const VueBinging = "Vue"
  const aliasHelper = (s: string) => `${helperMapName[s]}: _${helperMapName[s]}`
  push(`const { ${ast.helpers.map(aliasHelper)} } = ${VueBinging}; `)
  push("\n")
  push("return ")
}


function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterPolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ElEMENT:
      genElement(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
  }
}

function genCompoundExpression(node, context) {
  const children = node.children
  const { push } = context
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

//生成元素节点代码
function genElement(node, context) {
  const { push, helper } = context
  const { tag, children, props } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList(genNullable([tag, children, props]), context)
  push(')')
}

function genNodeList(nodes, context) {
  console.log(nodes)
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isString(node)) {
      push(node)
    } else {
      genNode(node, context)
    }

    if (i < nodes.length - 1) {
      push(", ")
    }
  }
}

function genNullable(args) {
  return args.map(arg => arg || "null")
}

function genExpression(node, context) {
  const { push } = context;
  push(`${node.content}`)
}

function genInterPolation(node, context) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

//生成文字节点代码
function genText(node, context) {
  const { push } = context
  push(`'${node.content}'`)
}

//创建上下文
function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context
}

