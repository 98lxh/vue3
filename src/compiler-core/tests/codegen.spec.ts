import { baseParse } from "../src/parse"
import { generate } from "../src/generate"
import { transform } from "../src/transform"

describe('codegen', () => {

  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    //快照测试
    expect(code).toMatchSnapshot(`return function render(_ctx,_cache,$props,$data,$options){
      return "hi"
    }`)
  })
})
