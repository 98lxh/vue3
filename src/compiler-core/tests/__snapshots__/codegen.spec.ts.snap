// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen string: return function render(_ctx,_cache,$props,$data,$options){
      return "hi"
    } 1`] = `"return function render(_ctx,_cache){return 'hi'}"`;
